contracts/mocks/tests/MathTest.sol

```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "forge-std/Test.sol";
+
+import "../../utils/math/Math.sol";
+import "../../utils/math/SafeMath.sol";
+
+contract MathTest is Test {
+    function testSquareRoot(uint256 input) public {
+        uint256 result = Math.sqrt(input);
+        uint256 square = result * result;
+
+        // `sqrt` should return a rounded-down value, so `square` is either less or equal to `input`.
+        if (square < input) {
+            // We know `result` is less than the true value for the square root, but we want to check that the error is
+            // minimal, in this case, less than 1.
+            // This should only happen if `input` is not a perfect square. We can then check that any result larger than
+            // the one we got would yield a square larger than `input`, meaning we got the smallest value for which the
+            // square is less than `input`.
+
+            (bool noOverflow, uint256 nextSquare) = SafeMath.tryMul(result + 1, result + 1);
```

Comment:
if `input` is `type(uint256).max`, and if `result` is correct, 

Then we have `result * result <= input` and `SafeMath.tryMul(result + 1, result + 1)` that overflow. The overflow here is correct.

IMO the check should possibly be

```
uint256 result = Math.sqrt(input);
uint256 resultSquare = result * result;
(bool noOverflow, uint256 resultPlusOneSquare) = SafeMath.tryMul(result + 1, result + 1);
asserTrue(resultSquare <= input && (!noOverflow || input < resultPlusOneSquare));
```

What do you think ?

Reply:
Yup, that seems fully equivalent. I didn't try very hard to have this be idiomatic or even not very verbose. I don't even have strong thoughts on what a good Solidity test might look like - if you think this reads better I can change it to look like that. I do find it slightly odd to run both assertions in the perfect square scenario though.